text/0058-valuecastable-format.md

```diff
@@ -0,0 +1,93 @@
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`ShapeCastable` subtypes can define their own ways of formatting values. For example, consider the following:
+
+```py
+class Opcode(amaranth.lib.enum.Enum, shape=unsigned(2)):
+    ADD = 0
+    SUB = 1
+    XOR = 2
+
+op = Signal(Opcode)
+m.d.comb += [
+    op.eq(Opcode.XOR)
+    Print(Format("current opcode: {}", op)),
+]
+```
+
+It is desirable that this example would print something like `"XOR"`, instead of `"2"`. Such functionality can be added on `ShapeCastable` subclasses by overriding the `format` method.
+
+However, sometimes it is also useful to print the raw value of such a signal. A shorthand is provided for that purpose:
+
+```py
+m.d.comb += Print(Format("current opcode: {op} (raw: {op!v})", op=op))
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new overridable method is added to `ShapeCastable`:
+
+- `format(self, value: ValueCastable, format_desc: str) -> Format`
+
+When a `ValueCastable` is formatted without specifying a conversion (ie. `"!r"` or `"!v"`):
+
+- `shape()` is called
+- if the shape is a plain `Shape`, `Value.cast()` is called on the value-castable, and formatting proceeds as if it was a plain value
+- otherwise (the shape is a `ShapeCastable`), the value-castable being formatted and the format descriptor (the part after `:` in `{}`, if any) are passed directly to `shape.format()`, and the result (which must be a `Format` object) is included directly in place of the format specifier
+
+The default implementation of `format` raises a `TypeError` exception.
+
+A new conversion, `!v`, is added to `Format` (with syntax analogous to `!r`). When specified, the value being formatted is converted through `Value.cast` before proceeding with further formatting. It can be used to print the raw value of value-castables. It is a no-op when used with a plain `Value`.
+
+An implementation of `__format__` is added to the `ValueCastable` base class that always raises a `TypeError`, directing the user to `Format` instead (like the one that already exists on plain `Value`).
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+None.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The `format` hook is added on `ShapeCastable` instead of `ValueCastable`. This ensures that `ValueCastable`s without a custom shape automatically get plain formatting.
+
+## Prior art
+[prior-art]: #prior-art
+
+This RFC is modelled directly on the `__format__` extension point in Python.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+What should be the default implementation of the `format` hook?
+
+The alternatives are:
+
+- (proposed above) raise `TypeError`
+- defer to plain value formatting
```

Comment:
I think this is probably the way to go. My reasoning is: imagine you are writing generic code, a FIFO or something. You want to display an element that is being popped. Your options:
- do not use `!v` and force consumers to sometimes implement a no-op `format` just so they can use your FIFO with debugging enabled;
- use `!v` and deny consumers which implement a special `format` legibility;
- branch on the exception or something (cursed, and probably what will actually take place eventually)

Reply:
This option is also the one which meshes best with our treatment of `ValueCastable`s otherwise (where they just decay to `Value`s unless something interferes with it).

Reply:
This is very convincing, I'll update the RFC.